crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +299,385 @@ where
         Ok(())
     }
 
+    /// Look up the IBC denomination from a IbcToken.
+    pub fn query_ibc_denom(
+        &self,
+        token: impl AsRef<str>,
+        owner: Option<&Address>,
+    ) -> Result<String> {
+        let hash = match Address::decode(token.as_ref()) {
+            Ok(Address::Internal(InternalAddress::IbcToken(hash))) => {
+                hash.to_string()
+            }
+            _ => return Ok(token.as_ref().to_string()),
+        };
+
+        if let Some(owner) = owner {
```

Comment:
Why do we need to know the owner to query the IBC denomination at all?

Reply:
I need to investigate this. For the purposes of validation, this function almost exactly mirrors the computation done during the construction of IBC transactions in the client. Compare this to https://github.com/anoma/namada/blob/08e56643fd65d42de3cf854122ab1fe74698f7ac/crates/sdk/src/rpc.rs#L1406 , which is invoked at https://github.com/anoma/namada/blob/08e56643fd65d42de3cf854122ab1fe74698f7ac/crates/sdk/src/tx.rs#L2583 . cc @yito88

Reply:
@yito88 can you clarify here? I'm quite confused about this

Reply:
`query_ibc_denom` is for looking for readable IBC denom with `IbcToken` hash.
Using an owner was to avoid looking up all IBC denoms when the balance query didn't require `--token`. (Now the query requires both the token and the owner. The RPC function is left for now.)
I think we don't need the query in MASP VP because the IBC denom should be in the packet and the token address can be calculated from it.

Reply:
For context, the token address for a receive packet is computed here in the MASP VP: https://github.com/anoma/namada/blob/4de6577c73e223504b0fcb226257ef7b85e248a7/crates/namada/src/ledger/native_vp/masp.rs#L483C25-L483C55 . This computation was designed to produce the same shielded token as the one produced by Hermes, which can be found at https://github.com/anoma/namada/blob/5cc23b4f43b12ecfc70eeca8d04cbe15f4b11a3a/crates/sdk/src/tx.rs#L3505 

> I think we don't need the query in MASP VP because the IBC denom should be in the packet and the token address can be calculated from it.

Thanks for the clarification @yito88 . Would this same reasoning apply to how Hermes derives the token from the IBC denom (also using `query_ibc_denom` and `received_ibc_token`)? Or is the context of the Hermes computation somehow different?

Reply:
`query_ibc_denom` is to look up for stored IBC denoms. `received_ibc_token` returns the IBC denom to be minted/unescrowed. Hermes doesn't derive the token.